snuba/query/expressions.py

```diff
@@ -0,0 +1,217 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import Callable, Iterator, Optional, Sequence
+
+from snuba.query.collections import NodeContainer
+
+
+@dataclass
+class Expression(ABC):
+    """
+    A node in the Query AST. This can be a leaf or an intermediate node.
+    It represents an expression that can be resolved to a value. This
+    includes column names, function calls and boolean conditions (which are
+    function calls themselves in the AST), literals, etc.
+
+    The root of the tree is not a Node itself yet (since it is the Query object).
+    Representing the root as a node itself does not seem very useful right now
+    since we never traverse the full tree. We could revisit that later.
+
+    All expression can have an optional alias.
+    """
+
+    alias: Optional[str]
+
+    @abstractmethod
+    def format(self) -> str:
+        """
+        Turn this node into a string for the Clickhouse query.
+
+        TODO: provide a clickhouse formatter to this method so that, through
+        a strategy pattern, this class will provide the content for the query
+        and the clickhouse formatter will provide the format.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def transform(self, func: Callable[[Expression], Expression]) -> Expression:
+        """
+        Transforms this expression through the function passed in input.
+        This works almost like a map function over sequences though, contrarily to
+        sequences, this acts on a subtree. The semantics of transform can be different
+        between intermediate nodes and leaves, so each node class can implement it
+        its own way.
+
+        It returns the new expression. It does not guarantee the returned value
+        will be a new object. No guarantee of immutability is given on expressions.
+        """
+        raise NotImplementedError
+
+
+class HierarchicalExpression(Expression, NodeContainer[Expression]):
```

Comment:
The only reason was that I did not like the idea of iterating on a leaf node (which does not contain nodes). Being able to do this seemed conceptually a counterintuitive api:
```
c = Column("alias", "column", "table")
for col in c:
    yield c
```

This is the sole reason for not making expression a NodeContainer itself.

About the OrderBy, the problem would exist anyway but it is not visible at this stage (mypy is fine now) because it would present itself when transforming a container of OrderBy classes. 
```
container: Sequence[OrderBy] = []
# stuff
container = map(o.transform(func), container)
```
This will happen no matter what in a different place (when I want to apply transformation across the board on the query). There there are two possible solution I am still thinking over:
- Do not transform across the board but let the user decide which parts of the query to transform (so the sequence of order by will take e more specialized mapping function. This require ORderBy to override transform
- When transforming the expressions across the board only iterate over the content of OrderBy, which is an Expression. Which is probably what most use case will want to.

Probably the Query api will need to expose the two approaches.

Reply:
I'll come back to this one, but I think my thoughts here relate to the other comment that I just made above in a way that I haven't had time to clearly articulate yet.

I do think that `OrderBy` is not an expression because it's not substitutable with other `Expression` types.

Reply:
Now the hierarchical node still exists but it is not needed for correctness, it is just used to reuse logic between FunctionCall and Aggregation. Everything is an Iterable